### PR TITLE
RPCN 0.9

### PR DESCRIPTION
**Pull request by RipleyTom**
### Added

- Presence support(should be easier to join games people in your friend list are playing)

### Fixed

- Public IP could be obtained from RPCN before being set
- Fixed SceNpTus* error checking

### Misc

- Updated Flatbuffers
- Reduced code repetition in rpcn_client.cpp

### CODE REVIEW

File: rpcs3/Emu/NP/np_helpers.cpp [hunk 1]
```diff
@@ -22,6 +22,34 @@ namespace np
...
+
+	template <typename T>
+	void onlinedata_to_presencedetails(const rpcn::friend_online_data& data, bool same_context, T& details)
+	{
+		memset(&details, 0, sizeof(T));
+		details.state = data.online ? (same_context ? SCE_NP_BASIC_PRESENCE_STATE_IN_CONTEXT : SCE_NP_BASIC_PRESENCE_STATE_OUT_OF_CONTEXT) : SCE_NP_BASIC_PRESENCE_STATE_OFFLINE;
+		strcpy_trunc(details.title, data.pr_title);
+		strcpy_trunc(details.status, data.pr_status);
+		strcpy_trunc(details.comment, data.pr_comment);
+		details.size = std::min<u32>(data.pr_data.size(), SCE_NP_BASIC_MAX_PRESENCE_SIZE);
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_helpers.cpp:41]:
```suggestion
		details.size = std::min<u32>(::size32(data.pr_data), SCE_NP_BASIC_MAX_PRESENCE_SIZE);
```

File: rpcs3/Emu/NP/rpcn_client.cpp [hunk 2]
```diff
@@ -2448,17 +2324,67 @@ namespace rpcn
...
+					presence_updates.insert_or_assign(username, u->second);
+				}
+
+				call_callbacks(ntype, username, online);
 			}
 			break;
 		}
+		case NotificationType::FriendPresenceChanged:
+		{
+			auto truncate_string = [](std::string& str, usz max_size)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/rpcn_client.cpp:2346]: considering that this function exists twice, you should probably move this to fmt::truncate.
Also, the `if (>=) resize(size-1)` logic can simply be written as `if (>) resize(size)`

File: rpcs3/Emu/NP/rpcn_client.cpp [hunk 3]
```diff
@@ -2607,4 +2533,25 @@ namespace rpcn
...
+		std::lock_guard lock(mutex_friends);
+
+		auto it = friend_infos.friends.begin();
+		while (it != friend_infos.friends.end() && index != 0)
+		{
+			it++;
+			index--;
+		}
+
+		return it == friend_infos.friends.end() ? std::nullopt : std::optional(*it);
```
Comment (Megamouse) [inline rpcs3/Emu/NP/rpcn_client.cpp:2547]: Can't this simply be written as:
```
if (index >= friend_infos.friends.size) return std::nullopt;
auto it = friend_infos.friends.begin();
std::advance(it, index);
return std::optional(*it);
```

File: rpcs3/Emu/NP/rpcn_client.cpp [hunk 4]
```diff
@@ -918,14 +945,38 @@ namespace rpcn
...
 		{
 			u32 num_friends = stream.get<u32>();
 			for (u32 i = 0; i < num_friends; i++)
 			{
 				std::string friend_name = stream.get_string(false);
-				bool online             = !!(stream.get<u8>());
-				friends.insert(std::make_pair(std::move(friend_name), std::make_pair(online, 0)));
+				bool online = !!(stream.get<u8>());
+
+				auto truncate_string = [](std::string& str, usz max_size)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/rpcn_client.cpp:956]: considering that this function exists twice, you should probably move this to fmt::truncate.
Also, the `if (>=) resize(size-1)` logic can simply be written as `if (>) resize(size)`

File: rpcs3/Emu/NP/np_handler.cpp [hunk 5]
```diff
@@ -793,10 +793,82 @@ namespace np
...
+			std::lock_guard lock(basic_handler.mutex);
+			memcpy(&basic_handler.context, context.get_ptr(), sizeof(basic_handler.context));
+			basic_handler.handler_func = handler;
+			basic_handler.handler_arg = arg;
+			basic_handler.context_sensitive = context_sensitive;
+			basic_handler_registered = true;
+
+			presence_self.pr_com_id = *context;
+			presence_self.pr_title = Emu.GetTitle();
+			if (presence_self.pr_title.size() >= SCE_NP_BASIC_PRESENCE_TITLE_SIZE_MAX)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.cpp:808]: See comment with fmt::truncate

File: rpcs3/Emu/NP/np_handler.cpp [hunk 6]
```diff
@@ -793,10 +793,82 @@ namespace np
...
+			{
+				presence_self.pr_title.resize(SCE_NP_BASIC_PRESENCE_TITLE_SIZE_MAX - 1);
+			}
+		}
+
+		if (g_cfg.net.psn_status != np_psn_status::psn_rpcn || !is_psn_active || !rpcn)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.cpp:814]: rpcn is checked twice

File: rpcs3/Emu/NP/np_handler.cpp [hunk 7]
```diff
@@ -1147,6 +1241,99 @@ namespace np
...
+		if (data)
+		{
+			if (data != presence_self.pr_data)
+			{
+				presence_self.pr_data = *data;
+				send_update = true;
+			}
+		}
+
+		if (send_update && is_psn_active && rpcn)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.cpp:1266]: rpcn checked twice?

File: rpcs3/Emu/NP/np_handler.cpp [hunk 8]
```diff
@@ -1147,6 +1241,99 @@ namespace np
...
+			}
+
+			rpcn->send_presence(presence_self.pr_com_id, presence_self.pr_title, presence_self.pr_status, presence_self.pr_comment, presence_self.pr_data);
+		}
+	}
+
+	template <typename T>
+	error_code np_handler::get_friend_presence_by_index(u32 index, SceNpUserInfo* user, T* pres)
+	{
+		if (!is_psn_active || g_cfg.net.psn_status != psn_rpcn || !rpcn)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.cpp:1281]: rpcn checked twice again, whereas user and pres aren't checked at all.

File: rpcs3/Emu/NP/np_handler.cpp [hunk 9]
```diff
@@ -1147,6 +1241,99 @@ namespace np
...
+		return CELL_OK;
+	}
+
+	template error_code np_handler::get_friend_presence_by_index<SceNpBasicPresenceDetails>(u32 index, SceNpUserInfo* user, SceNpBasicPresenceDetails* pres);
+	template error_code np_handler::get_friend_presence_by_index<SceNpBasicPresenceDetails2>(u32 index, SceNpUserInfo* user, SceNpBasicPresenceDetails2* pres);
+
+	template <typename T>
+	error_code np_handler::get_friend_presence_by_npid(const SceNpId& npid, T* pres)
+	{
+		if (!is_psn_active || g_cfg.net.psn_status != psn_rpcn || !rpcn)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.cpp:1311]: rpcn checked twice whereas pres isn't checked at all

File: rpcs3/Emu/NP/np_handler.cpp [hunk 10]
```diff
@@ -793,10 +793,82 @@ namespace np
 		return size_copied;
 	}
 
+	void np_handler::register_basic_handler(vm::cptr<SceNpCommunicationId> context, vm::ptr<SceNpBasicEventHandler> handler, vm::ptr<void> arg, bool context_sensitive)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.cpp:796]: no argument checks?

File: rpcs3/Emu/Cell/Modules/sceNp.cpp [hunk 11]
```diff
@@ -843,28 +835,27 @@ error_code sceNpBasicSetPresence(vm::cptr<u8> data, u32 size)
...
-	// {
-	// 	return SCE_NP_BASIC_ERROR_INVALID_ARGUMENT;
-	// }
-
 	if (size > SCE_NP_BASIC_MAX_PRESENCE_SIZE)
 	{
 		return SCE_NP_BASIC_ERROR_EXCEEDS_MAX;
 	}
 
+	// data is not checked
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/sceNp.cpp:848]: just ensure it then instead of provoking a crash 🤦 .

File: rpcs3/Emu/Cell/Modules/sceNp.cpp [hunk 12]
```diff
@@ -843,28 +835,27 @@ error_code sceNpBasicSetPresence(vm::cptr<u8> data, u32 size)
...
+
+	std::vector pr_data(data.get_ptr(), data.get_ptr() + size);
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/sceNp.cpp:850]: what if size is 0?

File: rpcs3/Emu/Cell/Modules/sceNp.cpp [hunk 13]
```diff
@@ -888,12 +879,29 @@ error_code sceNpBasicSetPresenceDetails(vm::cptr<SceNpBasicPresenceDetails> pres
 		return SCE_NP_BASIC_ERROR_EXCEEDS_MAX;
 	}
 
+	// pres is not checked
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/sceNp.cpp:882]: missing ensure where it's needed

File: rpcs3/Emu/Cell/Modules/sceNp.cpp [hunk 14]
```diff
@@ -917,6 +925,23 @@ error_code sceNpBasicSetPresenceDetails2(vm::cptr<SceNpBasicPresenceDetails2> pr
...
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/sceNp.cpp:928]: missing ensure where it's needed

File: rpcs3/Emu/Cell/Modules/sceNp.cpp (hunk 12 above)
Comment (RipleyTom) [reply rpcs3/Emu/Cell/Modules/sceNp.cpp:850]: it creates an empty vector?

Comment (elad335) [reply rpcs3/Emu/Cell/Modules/sceNp.cpp:850]: yeah

File: rpcs3/Emu/NP/rpcn_client.cpp [hunk 15]
```diff
@@ -1194,15 +1233,31 @@ namespace rpcn
...
+
+	std::map<std::string, friend_online_data> rpcn_client::get_presence_states()
+	{
+		std::lock_guard lock(mutex_friends);
+		return friend_infos.friends;
+	}
+
+	std::vector<u64> rpcn_client::get_new_messages()
+	{
+		std::vector<u64> ret_new_messages;
```
Comment (Megamouse) [inline rpcs3/Emu/NP/rpcn_client.cpp:1257]: Declaration can be moved to initialization

File: rpcs3/Emu/NP/rpcn_client.cpp [hunk 16]
```diff
@@ -1194,15 +1233,31 @@ namespace rpcn
 		return ret_replies;
 	}
 
-	std::vector<u64> rpcn_client::get_new_messages()
+	std::unordered_map<std::string, friend_online_data> rpcn_client::get_presence_updates()
```
Comment (Megamouse) [inline rpcs3/Emu/NP/rpcn_client.cpp:1236]: This function can be simplified just like the one below

File: rpcs3/Emu/NP/np_handler.cpp [hunk 17]
```diff
@@ -983,8 +1052,39 @@ namespace np
...
+						strcpy_trunc(to_add.from.name.data, npid);
+
+						if (std::memcmp(pr_info.pr_com_id.data, basic_handler.context.data, sizeof(pr_info.pr_com_id.data)) == 0)
+						{
+							to_add.event = SCE_NP_BASIC_EVENT_PRESENCE;
+							to_add.data = std::move(pr_info.pr_data);
+						}
+						else
+						{
+							if (basic_handler.context_sensitive && !pr_info.pr_com_id.data[0] == 0)
```
Comment (oltolm) [inline rpcs3/Emu/NP/np_handler.cpp:1071]:
```
[327/579] Building CXX object rpcs3/Emu/CMakeFiles/rpcs3_emu.dir/NP/np_handler.cpp.obj
C:/src/rpcs3/rpcs3/Emu/NP/np_handler.cpp:1076:47: warning: logical not is only applied to the left hand side of this comparison [-Wlogical-not-parentheses]
 1076 |                                                         if (basic_handler.context_sensitive && !pr_info.pr_com_id.data[0] == 0)
      |
 ^                          ~~
C:/src/rpcs3/rpcs3/Emu/NP/np_handler.cpp:1076:47: note: add parentheses after the '!' to evaluate the comparison first
 1076 |                                                         if (basic_handler.context_sensitive && !pr_info.pr_com_id.data[0] == 0)
      |
 ^
      |
  (                             )
C:/src/rpcs3/rpcs3/Emu/NP/np_handler.cpp:1076:47: note: add parentheses around left hand side expression to silence this warning
 1076 |                                                         if (basic_handler.context_sensitive && !pr_info.pr_com_id.data[0] == 0)
      |
 ^
      |
 (                         )
1 warning generated.
```

Comment (RipleyTom) [reply rpcs3/Emu/NP/np_handler.cpp:1071]: Thanks, will be fixed in https://github.com/RPCS3/rpcs3/pull/15025

**Pull request merged**